File: ios/Anemone SDK/model/generic/Address.swift
//
//  Address.swift
//  Anemone SDK
//
//  Created by Angel Garcia on 12/12/2016.
//  Copyright © 2016 Syltek Solutions S.L. All rights reserved.
//

import CoreLocation
import Foundation

public struct Address: Hashable {
    public let name: String?
    public let streetName: String?
    public let streetNumber: String?
    public let locality: String?
    public let subLocality: String?
    public let administrativeArea: String?
    public let subAdministrativeArea: String?
    public let postalCode: String?
    public let countryCode: String?
    public let countryName: String?
    public let coordinate: Coordinate?
    public let timeZone: TimeZone?

    public init(
        name: String? = nil,
        streetName: String? = nil,
        streetNumber: String? = nil,
        locality: String? = nil,
        subLocality: String? = nil,
        administrativeArea: String? = nil,
        subAdministrativeArea: String? = nil,
        postalCode: String? = nil,
        countryCode: String? = nil,
        countryName: String? = nil,
        coordinate: Coordinate? = nil,
        timeZone: TimeZone? = nil
    ) {
        self.name = name
        self.streetName = streetName
        self.streetNumber = streetNumber
        self.locality = locality
        self.subLocality = subLocality
        self.administrativeArea = administrativeArea
        self.subAdministrativeArea = subAdministrativeArea
        self.postalCode = postalCode
        self.countryCode = countryCode
        self.countryName = countryName
        self.coordinate = coordinate
        self.timeZone = timeZone
    }

    public func formattedShortAddress() -> String {
        if let name = name {
            return name
        }

        var addressComponents: [String] = []

        if let streetName = streetName {
            addressComponents.append(streetName)

            if let streetNumber = streetNumber {
                addressComponents.append(streetNumber)
            }
        }
        if let locality = locality {
            addressComponents.append(locality)
        }

        if addressComponents.isEmpty {
            return formattedDistrict()
        }

        addressComponents = addressComponents.filter { !$0.isEmpty }
        return addressComponents.joined(separator: ", ")
    }

    public func formattedFullAddress(useName: Bool = true) -> String {
        if let name = name, useName {
            return name
        }

        var addressComponents: [String] = []

        if let streetName = streetName {
            addressComponents.append(streetName)

            if let streetNumber = streetNumber {
                addressComponents.append(streetNumber)
            }
        }

        if let postalCode = postalCode {
            addressComponents.append(postalCode)
        }

        if let locality = locality {
            addressComponents.append(locality)
        }

        if let countryName = countryName {
            addressComponents.append(countryName)
        }

        addressComponents = addressComponents.filter { !$0.isEmpty }
        return addressComponents.joined(separator: ", ")
    }

    public func formattedDistrict() -> String {
        subLocality ?? locality ?? formattedRegion()
    }

    public func formattedRegion() -> String {
        var addressComponents: [String] = []

        if let administrativeArea = administrativeArea {
            addressComponents.append(administrativeArea)
        } else if let subAdministrativeArea = subAdministrativeArea {
            addressComponents.append(subAdministrativeArea)
        } else if let locality = locality {
            addressComponents.append(locality)
        }

        if let countryName = countryName {
            addressComponents.append(countryName)
        }

        addressComponents = addressComponents.filter { !$0.isEmpty }
        return addressComponents.joined(separator: ", ")
    }

    public func formattedLocality() -> String? {
        let components = [locality, subAdministrativeArea?.nonEmpty ?? administrativeArea]
            .compactMap { $0?.nonEmpty }
            .removingDuplicates()
        if components.count == 1 {
            return components[0]
        } else if components.count > 1 {
            return "\(components[0]) (\(components[1]))"
        } else {
            return nil
        }
    }
}

extension Address: JSONMappable {
    public init(json: JSONObject) throws {
        if json.has("html_attributions") || json.has("result") {
            try self.init(googlePlaceJson: try json.getJSONObject("result"))
        } else {
            try self.init(serverJSON: json)
        }
    }

    private init(serverJSON json: JSONObject) throws {
        self.streetName = json.optString("street")?.nonEmpty
        self.postalCode = json.optString("postal_code")?.nonEmpty
        self.locality = json.optString("city")?.nonEmpty
        self.coordinate = try? Coordinate.from(json: json.getJSONObject("coordinate"))
        self.name = json.optString("name")?.nonEmpty
        self.streetNumber = json.optString("street_number")?.nonEmpty
        self.subLocality = json.optString("sub_locality")?.nonEmpty
        self.administrativeArea = json.optString("administrative_area")?.nonEmpty
        self.subAdministrativeArea = json.optString("sub_administrative_area")?.nonEmpty
        self.countryCode = json.optString("country_code")?.nonEmpty
        self.countryName = json.optString("country")?.nonEmpty ?? json.optString("country_name")?.nonEmpty
        if let timezone = json.optString("timezone")?.nonEmpty {
            self.timeZone = TimeZone(identifier: timezone)
        } else {
            self.timeZone = nil
        }
    }

    // swiftlint:disable function_body_length
    private init(googlePlaceJson json: JSONObject) throws {
        self.name = try json.getString("name")
        let coordinateJson = try json.getJSONObject("geometry").getJSONObject("location")
        self.coordinate = Coordinate(lat: try coordinateJson.getDouble("lat"), lon: try coordinateJson.getDouble("lng"))

        var streetNumber: String?
        var streetName: String?
        var locality: String?
        var subLocality: String?
        var administrativeArea: String?
        var subAdministrativeArea: String?
        var postalCode: String?
        var countryName: String?
        var countryCode: String?

        let addressComponents = try json.getJSONArray("address_components").asJSONList()

        for component in addressComponents {
            guard let types = try? component.getStringArray("types") else { continue }
            let name = try? component.getString("long_name")

            if types.contains("street_number") {
                streetNumber = name
            } else if types.contains("route") {
                streetName = name
            } else if types.contains("locality") {
                locality = name
            } else if types.contains("sublocality") {
                subLocality = name
            } else if types.contains("administrative_area_level_1") {
                administrativeArea = name
            } else if types.contains("administrative_area_level_2") {
                subAdministrativeArea = name
            } else if types.contains("postal_code") {
                postalCode = name
            } else if types.contains("country") {
                countryName = name
                countryCode = try? component.getString("short_name")
            }
        }

        self.streetNumber = streetNumber
        self.streetName = streetName
        self.locality = locality
        self.subLocality = subLocality
        self.administrativeArea = administrativeArea
        self.subAdministrativeArea = subAdministrativeArea
        self.postalCode = postalCode
        self.countryName = countryName
        self.countryCode = countryCode
        self.timeZone = nil
    }
    // swiftlint:enable function_body_length
}

extension Address: JSONSerializable {
    public func toJson() -> JSONObject {
        let json = JSONObject()
        json.setString("street", streetName)
        json.setString("postal_code", postalCode)
        json.setString("city", locality)
        json.setObject("coordinate", coordinate?.toJson())
        json.setString("name", name)
        json.setString("street_number", streetNumber)
        json.setString("sub_locality", subLocality)
        json.setString("administrative_area", administrativeArea)
        json.setString("sub_administrative_area", subAdministrativeArea)
        json.setString("country_code", countryCode)
        json.setString("country", countryName)
        json.setString("timezone", timeZone?.identifier)

        return json
    }
}

public extension Address {
    func copy(
        streetName: String? = nil,
        timeZone: TimeZone? = nil
    ) -> Address {
        Address(
            name: name,
            streetName: streetName ?? self.streetName,
            streetNumber: streetNumber,
            locality: locality,
            subLocality: subLocality,
            administrativeArea: administrativeArea,
            subAdministrativeArea: subAdministrativeArea,
            postalCode: postalCode,
            countryCode: countryCode,
            countryName: countryName,
            coordinate: coordinate,
            timeZone: timeZone ?? self.timeZone
        )
    }
}

public extension Address {
    init(_ placemark: CLPlacemark, coordinate: Coordinate?) {
        self.name = placemark.name
        self.streetName = placemark.thoroughfare
        self.streetNumber = placemark.subThoroughfare
        self.locality = placemark.locality
        self.subLocality = placemark.subLocality
        self.administrativeArea = placemark.administrativeArea
        self.subAdministrativeArea = placemark.subAdministrativeArea
        self.postalCode = placemark.postalCode
        self.countryCode = placemark.isoCountryCode
        self.countryName = placemark.country
        self.timeZone = placemark.timeZone

        if let coordinate = coordinate {
            self.coordinate = coordinate
        } else if let location = placemark.location {
            self.coordinate = Coordinate(location.coordinate)
        } else {
            self.coordinate = nil
        }
    }
}

extension Address: Equatable { }
public func == (lhs: Address, rhs: Address) -> Bool {
    lhs.coordinate == rhs.coordinate
}

public extension Address {
    func locationText() -> String? {
        var text: String? = nil
        allNotNull(self.locality, self.countryName)
            .also { city, country in
                text = "\(city), \(country)"
            }
        return text
    }
}

File: android/app/src/main/java/com/playtomic/test/anemone_sdk/model/generic/Address.kt
package com.anemonesdk.model.generic

import android.text.TextUtils
import com.anemonesdk.general.extension.from
import com.anemonesdk.general.extension.toJson
import com.anemonesdk.general.json.JSONMappable
import com.anemonesdk.general.json.JSONObject
import com.playtomic.foundation.extension.allNotNull
import com.playtomic.foundation.extension.compactMap
import com.playtomic.foundation.extension.nonEmpty
import com.playtomic.foundation.extension.removingDuplicates
import com.playtomic.foundation.extension.tryOrNull
import com.playtomic.foundation.model.Coordinate
import org.json.JSONException
import java.util.*

/**
 * Created by agarcia on 23/12/2016.
 */

class Address : JSONMappable {

    var name: String? = null

    var streetName: String? = null

    var streetNumber: String? = null

    var locality: String? = null

    var subLocality: String? = null

    var administrativeArea: String? = null

    var subAdministrativeArea: String? = null

    var postalCode: String? = null

    var countryCode: String? = null

    var countryName: String? = null

    var coordinate: Coordinate? = null

    var timeZone: TimeZone? = null

    constructor(coordinate: Coordinate) {
        this.coordinate = coordinate
    }

    constructor(name: String, coordinate: Coordinate) {
        this.name = name
        this.coordinate = coordinate
    }

    constructor(address: android.location.Address, coordinate: Coordinate?) {
        this.name = address.featureName
        this.streetName = address.thoroughfare
        this.streetNumber = address.subThoroughfare
        this.locality = address.locality
        this.subLocality = address.subLocality
        this.administrativeArea = address.adminArea
        this.subAdministrativeArea = address.subAdminArea
        this.postalCode = address.postalCode
        this.countryCode = address.countryCode
        this.countryName = address.countryName
        if (coordinate == null && address.hasLatitude() && address.hasLongitude()) {
            this.coordinate = Coordinate(address.latitude, address.longitude)
        } else {
            this.coordinate = coordinate
        }

        // Not nice but sometimes google api returns same info in streetNumber and name. Name in that case is wrong and should include the streetName
        if (name != null && streetNumber != null && streetName != null && name == streetNumber) {
            this.name = streetName + ", " + streetNumber
        }
    }

    @Throws(JSONException::class)
    constructor(json: JSONObject) : super(json) {
        if (json.has("html_attributions") || json.has("result")) {
            mapFromGooglePlacesJSON(json.getJSONObject("result"))
        } else {
            mapFromServerJSON(json)
        }
    }

    constructor(
        name: String? = null,
        streetName: String? = null,
        streetNumber: String? = null,
        locality: String? = null,
        subLocality: String? = null,
        administrativeArea: String? = null,
        subAdministrativeArea: String? = null,
        postalCode: String? = null,
        countryCode: String? = null,
        countryName: String? = null,
        coordinate: Coordinate? = null,
        timeZone: TimeZone? = null
    ) : super() {
        this.name = name
        this.streetName = streetName
        this.streetNumber = streetNumber
        this.locality = locality
        this.subLocality = subLocality
        this.administrativeArea = administrativeArea
        this.subAdministrativeArea = subAdministrativeArea
        this.postalCode = postalCode
        this.countryCode = countryCode
        this.countryName = countryName
        this.coordinate = coordinate
        this.timeZone = timeZone
    }

    fun formattedShortAddress(): String {
        name?.let {
            return it
        }

        var addressComponents: MutableList<String> = ArrayList()
        streetName?.let { streetName ->
            addressComponents.add(streetName)

            streetNumber?.let { streetNumber ->
                addressComponents.add(streetNumber)
            }
        }
        locality?.let { locality ->
            addressComponents.add(locality)
        }

        if (addressComponents.isEmpty()) {

            return formattedDistrict()
        }

        addressComponents = addressComponents.filter { !TextUtils.isEmpty(it) }.toMutableList()
        return TextUtils.join(", ", addressComponents)
    }

    fun formattedFullAddress(useName: Boolean = true): String {
        val currentName = name
        if (useName && currentName != null) return currentName
        var addressComponents: MutableList<String> = ArrayList()

        streetName?.let { streetName ->
            addressComponents.add(streetName)

            streetNumber?.let { streetNumber ->
                addressComponents.add(streetNumber)
            }
        }

        postalCode?.let { postalCode ->
            addressComponents.add(postalCode)
        }

        locality?.let { locality ->
            addressComponents.add(locality)
        }

        countryName?.let { countryName ->
            addressComponents.add(countryName)
        }

        addressComponents = addressComponents.filter { !TextUtils.isEmpty(it) }.toMutableList()
        return TextUtils.join(", ", addressComponents)
    }

    fun formattedDistrict(): String =
        subLocality ?: locality ?: formattedRegion()

    fun formattedRegion(): String {
        var addressComponents: MutableList<String> = ArrayList()

        val administrativeArea = administrativeArea
        val subAdministrativeArea = subAdministrativeArea
        val locality = locality
        if (administrativeArea != null) {
            addressComponents.add(administrativeArea)
        } else if (subAdministrativeArea != null) {
            addressComponents.add(subAdministrativeArea)
        } else if (locality != null) {
            addressComponents.add(locality)
        }

        countryName?.let { countryName ->
            addressComponents.add(countryName)
        }

        addressComponents = addressComponents.filter { !TextUtils.isEmpty(it) }.toMutableList()
        return TextUtils.join(", ", addressComponents)
    }

    fun formattedLocation(): String? {
        val cityComponents = listOf(locality, subAdministrativeArea?.nonEmpty ?: administrativeArea)
            .compactMap { it?.nonEmpty }
            .removingDuplicates()
        val cityString = if (cityComponents.count() == 1) {
            cityComponents[0]
        } else if (cityComponents.size > 1) {
            "${cityComponents[0]}, ${cityComponents[1]}"
        } else {
            null
        }
        return allNotNull(cityString, countryName)?.let {
            "${it.value1}, ${it.value2}"
        }
    }

    fun formattedLocality(): String? {
        val components = listOf(locality, subAdministrativeArea?.nonEmpty ?: administrativeArea)
            .compactMap { it?.nonEmpty }
            .removingDuplicates()
        if (components.count() == 1) {
            return components[0]
        } else if (components.size > 1) {
            return "${components[0]} (${components[1]})"
        } else {
            return null
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val address = other as Address?

        return if (coordinate != null) coordinate == address?.coordinate else address?.coordinate == null
    }

    override fun hashCode(): Int {
        return coordinate?.hashCode() ?: 0
    }

    @Throws(JSONException::class)
    private fun mapFromServerJSON(json: JSONObject) {
        streetName = json.optString("street")?.nonEmpty
        postalCode = json.optString("postal_code")?.nonEmpty
        locality = json.optString("city")?.nonEmpty
        coordinate = tryOrNull { Coordinate.from(json.getJSONObject("coordinate")) }
        name = json.optString("name")?.nonEmpty
        streetNumber = json.optString("street_number")?.nonEmpty
        subLocality = json.optString("sub_locality")?.nonEmpty
        administrativeArea = json.optString("administrative_area")?.nonEmpty
        subAdministrativeArea = json.optString("sub_administrative_area")?.nonEmpty
        countryCode = json.optString("country_code")?.nonEmpty
        countryName = json.optString("country")?.nonEmpty ?: json.optString("country_name")?.nonEmpty
        val timeZone = json.optString("timezone")?.nonEmpty
        if (timeZone != null) {
            this.timeZone = TimeZone.getTimeZone(timeZone)
        }
    }

    @Throws(JSONException::class)
    private fun mapFromGooglePlacesJSON(json: JSONObject) {
        name = json.getString("name")
        val coordinateJson = json.getJSONObject("geometry").getJSONObject("location")
        coordinate = Coordinate(coordinateJson.getDouble("lat"), coordinateJson.getDouble("lng"))

        val addressComponents = json.getJSONArray("address_components").asJSONList()
        for (component in addressComponents) {
            val types = component.optStringArray("types")
            val name = component.optString("long_name")
            if (types != null) {

                if (types.contains("street_number")) {
                    streetNumber = name
                } else if (types.contains("route")) {
                    streetName = name
                } else if (types.contains("locality")) {
                    locality = name
                } else if (types.contains("sublocality")) {
                    subLocality = name
                } else if (types.contains("administrative_area_level_1")) {

                    administrativeArea = name
                } else if (types.contains("administrative_area_level_2")) {
                    subAdministrativeArea = name
                } else if (types.contains("postal_code")) {
                    postalCode = name
                } else if (types.contains("country")) {
                    countryName = name
                    countryCode = component.optString("short_name")
                }
            }
        }
    }
}

fun Address?.locationText(): String {
    return allNotNull(this?.locality, this?.countryName)?.let { "${it.value1}, ${it.value2}" } ?: ""
}
